### PR TITLE
Upgraded versions for angular, typescript, tslint and their dependencies

### PR DESCRIPTION
**Pull request by damianr13**

### CODE REVIEW

**Comment by damianr13** [comment]
This pull request was updated with solution for #205

**Comment by damianr13** [comment]
Updated with solution for #187

**Comment by damianr13** [comment]
Updated with partial solution for #184

**Comment by aniri** [comment]
@idormenco could you please review this? :smile:

**Comment by idormenco** [comment]
@aniri yes, i will

File: frontend/src/app/routing/guards/load-answer-list.guard.ts [hunk 1]
```diff
@@ -9,13 +11,13 @@ export class AnswerListGuard implements CanActivate {
...
-            .map(s => new LoadAnswerPreviewAction(route.data['urgent'], 1, 5, true, s))
-            .map(a => {
-                this.store.dispatch(a)
-            })
-            .subscribe()
+        this.store.select(s => s.answer.answerFilters).pipe(take(1),
+            map(s => new LoadAnswerPreviewAction(route.data['urgent'], 1, 5, true, s)),
+            map(a => {
+                this.store.dispatch(a);
+            }), )
```
Comment (bogdanconstantinescu) [inline frontend/src/app/routing/guards/load-answer-list.guard.ts]: `this.store.PIPE(select(), take(), map(), map())`

File: frontend/src/app/routing/guards/load-statistics.guard.ts [hunk 2]
```diff
@@ -15,22 +18,22 @@ export class LoadStatisticsGuard implements CanActivate {
     canActivate(route: ActivatedRouteSnapshot, state: RouterStateSnapshot) {
         if (route.params['key']) {
             this.store
-                .select(s => s.statistics[route.params['key']])
-                .take(1)
-                .map(s => new LoadStatisticAction(s.key, 1, 20, true))
+                .select(s => s.statistics[route.params['key']]).pipe(
+                take(1),
+                map(s => new LoadStatisticAction(s.key, 1, 20, true)), )
```
Comment (bogdanconstantinescu) [inline frontend/src/app/routing/guards/load-statistics.guard.ts]: `this.store.PIPE(select(), take(), map())`

File: frontend/src/app/routing/guards/load-statistics.guard.ts [hunk 3]
```diff
@@ -15,22 +18,22 @@ export class LoadStatisticsGuard implements CanActivate {
...
+                .select(s => s.statistics).pipe(
+                take(1),
                 // .do(console.log)
-                .map(s => values(s))
-                .concatMap(s => Observable.from(s))
+                map(s => values(s)),
+                concatMap(s => observableFrom(s)),
                 // .do(console.log)
-                .map((i: StatisticsStateItem) => new LoadStatisticAction(i.key, 1, 5, true))
+                map((i: StatisticsStateItem) => new LoadStatisticAction(i.key, 1, 5, true)), )
```
Comment (bogdanconstantinescu) [inline frontend/src/app/routing/guards/load-statistics.guard.ts]: `this.store.PIPE(select(), take(), map(), concatMap(), map())`

File: frontend/src/app/store/answer/answer.effects.ts [hunk 4]
```diff
@@ -39,9 +42,9 @@ export class AnswerEffects {
 
   @Effect()
   loadThreads = this.actions
-    .ofType(AnswerActionTypes.LOAD_PREVIEW)
-    .filter((a: LoadAnswerPreviewAction) => shouldLoadPage(a.payload.page, a.payload.pageSize, this.state.threads.length))
-    .switchMap((action: LoadAnswerPreviewAction) => {
+    .pipe(ofType(AnswerActionTypes.LOAD_PREVIEW)).pipe(
```
Comment (bogdanconstantinescu) [inline frontend/src/app/store/answer/answer.effects.ts]: 2 `.pipe()`s are unnecessary - just use one: `.pipe(ofType(), filter(), switchMap())`

File: frontend/src/app/store/answer/answer.effects.ts [hunk 5]
```diff
@@ -80,8 +83,8 @@ export class AnswerEffects {
 
   @Effect()
   loadDetails = this.actions
-    .ofType(AnswerActionTypes.LOAD_DETAILS)
-    .switchMap((action: LoadAnswerDetailsAction) => {
+    .pipe(ofType(AnswerActionTypes.LOAD_DETAILS)).pipe(
+    switchMap((action: LoadAnswerDetailsAction) => {
```
Comment (bogdanconstantinescu) [inline frontend/src/app/store/answer/answer.effects.ts]: Same

File: frontend/src/app/store/answer/answer.effects.ts [hunk 6]
```diff
@@ -91,25 +94,25 @@ export class AnswerEffects {
...
-    .ofType(AnswerActionTypes.LOAD_DETAILS)
-    .map((a: LoadAnswerDetailsAction) => new LoadNotesAction(a.payload.sectionId, a.payload.observerId));
+    .pipe(ofType(AnswerActionTypes.LOAD_DETAILS)).pipe(
+    map((a: LoadAnswerDetailsAction) => new LoadNotesAction(a.payload.sectionId, a.payload.observerId)));
 
   @Effect()
   loadExtraFromAnswer = this.actions
-    .ofType(AnswerActionTypes.LOAD_DETAILS)
-    .map((a: LoadAnswerDetailsAction) => new LoadAnswerExtraAction(a.payload.observerId, a.payload.sectionId));
+    .pipe(ofType(AnswerActionTypes.LOAD_DETAILS)).pipe(
```
Comment (bogdanconstantinescu) [inline frontend/src/app/store/answer/answer.effects.ts]: Same

File: frontend/src/app/store/form/form.effects.ts [hunk 7]
```diff
@@ -19,27 +22,27 @@ export class FormEffects {
...
-        .switchMap(r=>r.formVersions)
-        .map(f=>{
-         return { id:f.id,description: f.description};
-        })
-        .concatMap((x: { id: number, description: string }) => this.getForm(x.id, x.description))
-        .map(form => new FormLoadCompletedAction([form]))
-        .catch(() => Observable.of(new FormErrorAction()));
-
-    private getForm(id: number,description: string): Observable<Form> {
+        .pipe(ofType(FormActionTypes.LOAD)).pipe(
```
Comment (bogdanconstantinescu) [inline frontend/src/app/store/form/form.effects.ts]: Only one `.pipe()` please

File: frontend/src/app/store/note/note.effects.ts [hunk 8]
```diff
@@ -16,11 +18,11 @@ export class NoteEffects {
 
     @Effect()
     notesStream = this.actions
-        .ofType(NoteActionTypes.LOAD)
-        .switchMap((a: LoadNotesAction) => {
+        .pipe(ofType(NoteActionTypes.LOAD)).pipe(
```
Comment (bogdanconstantinescu) [inline frontend/src/app/store/note/note.effects.ts]: Only one `.pipe()` please

File: frontend/src/app/store/observers/observers.effects.ts [hunk 9]
```diff
@@ -26,65 +29,65 @@ export class ObserversEffects {
   private baseUrl: string;
 
   constructor(private http: ApiService, private actions: Actions, private store: Store<AppState>) {
-    store.select(s => s.observers).subscribe(s => this.state = s)
+    store.select(s => s.observers).subscribe(s => this.state = s);
```
Comment (bogdanconstantinescu) [inline frontend/src/app/store/observers/observers.effects.ts]: `this.store.PIPE(select())`

File: frontend/src/app/store/observers/observers.effects.ts [hunk 10]
```diff
@@ -26,65 +29,65 @@ export class ObserversEffects {
...
 
   @Effect()
   loadStats = this.actions
-    .ofType(ObserversActions.LOAD)
-    .map(a => <LoadObserversAction>a)
-    .filter((a: LoadObserversAction) => shouldLoadPage(a.payload.page, a.payload.pageSize, this.state[a.payload.key].values.length))
-    .groupBy(a => a.payload.key)
-    .flatMap((obs) =>
-      obs.switchMap((a) => {
+    .pipe(ofType(ObserversActions.LOAD)).pipe(
```
Comment (bogdanconstantinescu) [inline frontend/src/app/store/observers/observers.effects.ts]: Only one `.pipe()` please

File: frontend/src/app/store/observers/observers.effects.ts [hunk 11]
```diff
@@ -26,65 +29,65 @@ export class ObserversEffects {
...
 
 
   @Effect()
   deleteObserver = this.actions
-    .ofType(ObserversActions.DELETE)
-    .map(a => <DeleteObserverAction>a)
-    .groupBy(a => a.payload.key)
-    .flatMap((obs) =>
-      obs.switchMap((a) => {
+    .pipe(ofType(ObserversActions.DELETE)).pipe(
```
Comment (bogdanconstantinescu) [inline frontend/src/app/store/observers/observers.effects.ts]: Same

File: frontend/src/app/store/observers/observers.effects.ts [hunk 12]
```diff
@@ -93,23 +96,23 @@ export class ObserversCountEffects {
   private baseUrl: string;
 
   constructor(private http: ApiService, private actions: Actions, private store: Store<AppState>) {
-    store.select(s => s.observersCount).subscribe(s => this.state = s)
+    store.select(s => s.observersCount).subscribe(s => this.state = s);
```
Comment (bogdanconstantinescu) [inline frontend/src/app/store/observers/observers.effects.ts]: `this.store.PIPE(select())`

File: frontend/src/app/store/observers/observers.effects.ts [hunk 13]
```diff
@@ -93,23 +96,23 @@ export class ObserversCountEffects {
...
     this.baseUrl = environment.apiUrl;
   }
 
 
   @Effect()
   loadStats = this.actions
-    .ofType(ObserversActions.LOADOBSERVERSTOTALCOUNT)
-    .switchMap((obs) =>{
+    .pipe(ofType(ObserversActions.LOADOBSERVERSTOTALCOUNT)).pipe(
```
Comment (bogdanconstantinescu) [inline frontend/src/app/store/observers/observers.effects.ts]: Only one `.pipe()` please

File: frontend/src/app/store/statistics/statistics.effects.ts [hunk 14]
```diff
@@ -24,12 +26,12 @@ export class StatisticsEffects {
 
     @Effect()
     loadStats = this.actions
-        .ofType(StatisticsActions.LOAD)
-        .map(a => <LoadStatisticAction>a)
-        .filter((a: LoadStatisticAction) => shouldLoadPage(a.payload.page, a.payload.pageSize, this.state[a.payload.key].values.length))
-        .groupBy(a => a.payload.key)
-        .flatMap((obs) =>
-            obs.switchMap((a) => {
+        .pipe(ofType(StatisticsActions.LOAD)).pipe(
```
Comment (bogdanconstantinescu) [inline frontend/src/app/store/statistics/statistics.effects.ts]: Only one `.pipe()` please

File: frontend/src/app/store/statistics/statistics.state.ts [hunk 15]
```diff
@@ -4,31 +4,31 @@ import { LabelValueModel } from '../../models/labelValue.model';
...
+            this.subHeader = config.subHeader;
         }
     }
 }
 export class StatisticsState {
     [key: string]: StatisticsStateItem
 }
-export let statisticsInitialState: StatisticsState = _.keyBy<StatisticsStateItem>(statisticsConfig.map<StatisticsStateItem>((config) => new StatisticsStateItem(config)),value => value.key);
\ No newline at end of file
+export let statisticsInitialState: StatisticsState = _.keyBy<StatisticsStateItem>(statisticsConfig.map<StatisticsStateItem>((config) => new StatisticsStateItem(config)), value => value.key);
```
Comment (bogdanconstantinescu) [inline frontend/src/app/store/statistics/statistics.state.ts]: Just import what you need from lodash, not the whole library

File: frontend/src/app/store/store.module.ts [hunk 16]
```diff
@@ -51,15 +53,15 @@ if (!environment.production) {
 export class AppStoreModule {
     constructor(store: Store<AppState>, tokenService: TokenService) {
         tokenService.tokenStream.subscribe((token) => {
-            let clearForms = !token;
-            store.select(s => s.form).take(1).subscribe(s => {
+            const clearForms = !token;
+            store.select(s => s.form).pipe(take(1)).subscribe(s => {
```
Comment (bogdanconstantinescu) [inline frontend/src/app/store/store.module.ts]: `this.store.PIPE(select(), take())`

**Comment by aniri** [comment]
@damianr13 hey :) could you please also fix the merge conflicts? thanks!

**Comment by damianr13** [comment]
There should be no more conflicts but now it needs to be reviewed again @aniri  @bogdanconstantinescu

**Pull request merged**